ferrocene/ci/scripts/llvm-cache.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: MIT OR Apache-2.0
+# SPDX-FileCopyrightText: The Ferrocene Developers
+
+import argparse
+import subprocess
+import hashlib
+import os
+import io
+import shutil
+import pathlib
+import tarfile
+import sys
+
+CACHE_BUCKET="ferrocene-ci-caches"
+CACHE_PREFIX="prebuilt-llvm"
+
+
+TARBALL_PATH = "llvm-cache.tar"
+COMPRESSED_TARBALL_PATH = f"{TARBALL_PATH}.zst"
+
+# Note that this *ignores* symlinks. If you need a binary that's actually a
+# symlink please add to the list the binary the symlink points *to*.
+KEEP_LLVM_BINARIES=[
+    # Needed for the `llvm-tools` component
+    "llvm-cov",
+    "llvm-nm",
+    "llvm-objcopy",
+    "llvm-objdump",
+    "llvm-profdata",
+    "llvm-readobj",
+    "llvm-size",
+    "llvm-strip",
+    "llvm-ar",
+    "llvm-as",
+    "llvm-dis",
+    "llc",
+    "opt",
+    # Used by `llvm-ar`
+    "llvm-ranlib",
+    "llvm-dlltool",
+    "llvm-lib",
+    # Used by `llvm-objdump`
+    "llvm-otool",
+    # Used by `llvm-objcopy`, `llvm-strip`
+    "llvm-bitcode-strip",
+    # Used by `llvm-readobj`
+    "llvm-readelf",
+    # Needed to link `rustc` with LLVM
+    "llvm-config",
+    # Needed by the Rust test suite
+    "llvm-bcanalyzer",
+    "llvm-dwarfdump",
+    "FileCheck",
+    # Needed to resolve symbols
+    "llvm-dwp",
+    # Needed to build LLD
+    "llvm-tblgen",
+]
+
+def arguments():
+    parser = argparse.ArgumentParser(
+        prog="llvm-cache.py",
+        description="Report various data about LLVM caches",
+    )
+    subparsers = parser.add_subparsers(dest="subcommand", help="sub-command help")
+
+    prepare_parser = subparsers.add_parser("prepare", help="Build and cache LLVM")
+    
+    download_parser = subparsers.add_parser("download", help="Download the existing LLVM cache")
+
+    s3_url_parser = subparsers.add_parser("s3-url", help="Calculate the LLVM cache URL")
+
+    return parser.parse_args()
+
+
+def run():
+    args = arguments()
+    try:
+        ferrocene_host = os.environ["FERROCENE_HOST"]
+    except:
+        print("Set FERROCENE_HOST environment to a Rust triple")
+        exit(1)
+
+    # match added in 3.10
+    if args.subcommand == "s3-url":
+        subcommand_s3_url(ferrocene_host)
+    elif args.subcommand == "download":
+        subcommand_download(ferrocene_host)
+    elif args.subcommand == "prepare":
+        subcommand_prepare(ferrocene_host)
+    else:
+        print(f"Unknown command {args.subcommand}")
+
+def subcommand_download(ferrocene_host):
+    s3_url = get_s3_url(ferrocene_host)
+
+    s3_cp_cmd = shlex.quote(f"aws s3 cp {s3_url} - | zstd --decompress - -o {TARBALL_PATH}")
```

Comment:
The individual arguments should be wrapped in shlex.quote:

```suggestion
    s3_cp_cmd = f"aws s3 cp {shlex.quote(s3_url)} - | zstd --decompress - -o {shlex.quote(TARBALL_PATH)}"
```

Reply:
Oh my gosh, I can't believe I missed that. Whoops.